fix: empty buckets/objects nodes in new setup

anjalshireesh opened the pull request:
## Description

When the number of buckets/objects is zero, it gets omitted from the
health report json as `0` is considered an empty value by the json
marshaller.

Remove `omitempty` from the bucket/object count fields and add an
`error` field. When the count cannot be fetched because of any error,
the `error` field will be populated with the actual error message. (Note
that the count will also get populated with the default value of `0`).
  
The client should read the `count` value only if there is no error.

The new behavior will be as follows:

- non-zero (say 3) buckets

```json
"buckets": {
  "count": 3
},
```

- no buckets

```json
"buckets": {
  "count": 0
},
```

- bucket count cannot be fetched because of an error

```json
"buckets": {
  "count": 0,
  "error": "error message"
},
```

## Motivation and Context

Bugfix

## How to test this PR?

- Create a new minio cluster
- Build `mc` with dependency on this PR
- Generate the health report for this cluster (`./mc admin subnet health {alias]`)
- Check that the generated report contains values for `count` under `buckets` and `objects` as `0`
- Generate the report using a previous version of mc
- Verify that the report generation command doesn't fail, and that the `buckets` and `objects` nodes are empty (`{}`)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

anjalshireesh committed:
fix: empty buckets/objects nodes in new setup

When the number of buckets/objects is zero, it gets omitted from the
health report json as `0` is considered an empty value by the json
marshaller.

Changing the type of these fields to pointer resolves this and helps in
clearly distinguishing between zero value and no data.

klauspost commented on pkg/madmin/info-commands.go:
```diff
@@ -195,17 +195,17 @@ type Services struct {
 // Buckets contains the number of buckets
 type Buckets struct {
-	Count uint64 `json:"count,omitempty"`
+	Count *uint64 `json:"count,omitempty"`
```
Why not just remove the omitempty instead?

anjalshireesh replied:
The values are fetched using the function `loadDataUsageFromBackend` and populated in the struct only when this function doesn't return an error. In case of error, the count cannot be set.

When there are actually no buckets, we want

```json
"buckets": {
  "count": 0
},
```
when the bucket count could not be fetched because of some error, we want
```json
"buckets": {
},
```

harshavardhana replied:
Using pointers for integers is a very non idiomatic style for Go, if needed you should have an error string additionally that provides this detail or use UintMax upon error as an unexpected value.

anjalshireesh replied:
Will look into removing the `omitempty` + adding `Error` field

anjalshireesh replied:
@klauspost @harshavardhana I've updated the PR with following changes:

 - removed `omitempty` from the `count` fields
 - added an `error` field in each of these `count` fields
 - removed usage of pointers. Client can now rely on the `error` field to know if there was any error when fetching this data.
 
 PTAL

anjalshireesh committed:
Add error field instead of using pointers

Remove `omitempty` from the bucket/object count fields and add an
`error` field. When the count cannot be fetched because of any error,
the `error` field will be populated with the actual error message. (Note
that the count will also get populated with the default value of `0`).

The client should read the `count` value only if there is no error.
